src/v/rpc/connection_cache.h

```diff
@@ -24,10 +24,62 @@
 #include <seastar/core/sharded.hh>
 #include <seastar/core/shared_ptr.hh>
 
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+
 #include <chrono>
 #include <unordered_map>
 
 namespace rpc {
+
+class connection_allocation_strategy {
+public:
+    connection_allocation_strategy(
+      unsigned max_connections_per_node, unsigned total_shards);
+
+    // \brief Returns whether a node has already been assigned connections.
+    bool has_connection_assignments_for(model::node_id) const;
+
+    struct connection_assignment_action {
+        model::node_id node;
+        ss::shard_id shard;
+    };
+    using remove_mapping_action = connection_assignment_action;
+
+    struct shard_assignment_action {
+        model::node_id node;
+        ss::shard_id src_shard;
+        ss::shard_id conn_shard;
+    };
+
+    struct changes {
+        std::vector<connection_assignment_action> add_connections;
+        std::vector<connection_assignment_action> remove_connections;
+        std::vector<shard_assignment_action> add_mapping;
+        std::vector<remove_mapping_action> remove_mapping;
+    };
+
+    changes create_connection_assignments_for(model::node_id);
+    changes remove_connection_assignments_for(model::node_id);
+
+    const absl::flat_hash_set<ss::shard_id>&
+    get_connection_shards(model::node_id id) const {
+        return _node_to_shards.at(id);
+    }
+
+private:
+    unsigned _total_shards;
+    unsigned _max_connections_per_node;
+
+    absl::flat_hash_map<model::node_id, absl::flat_hash_set<ss::shard_id>>
+      _node_to_shards;
+    std::vector<std::pair<ss::shard_id, size_t>> _connections_per_shard;
```

Comment:
Using a struct would be nice here as well for readability as it avoids the std::get stuff.